File: solutions/longest_substring/longest_substring.cpp
#include <bits/stdc++.h>
using namespace std;

int lengthOfLongestSubstring(string s) {
    unordered_map<char, int> charIndex;  // store last seen index of each character
    int maxLength = 0;
    int start = 0;  // left pointer of the window

    for (int end = 0; end < s.size(); end++) {
        char currentChar = s[end];

        // If we have seen this char before and it’s inside the current window
        if (charIndex.find(currentChar) != charIndex.end() && charIndex[currentChar] >= start) {
            start = charIndex[currentChar] + 1; // move start to right of the duplicate
        }

        charIndex[currentChar] = end; // update last seen index
        maxLength = max(maxLength, end - start + 1); // update window size
    }

    return maxLength;
}

int main() {
    string s;
    cout << "Enter a string: ";
    cin >> s;

    cout << "Length of longest substring without repeating characters: "
         << lengthOfLongestSubstring(s) << endl;

    return 0;
}

File: solutions/longest_substring/longest_substring.py
def longestSubstring(s):
    mapping = {}  # stores last index of each character
    start = 0     # start of the current window
    max_length = 0

    for end in range(len(s)):
        char = s[end]

        # if the character is already seen and inside the current window
        if char in mapping and mapping[char] >= start:
            start = mapping[char] + 1  # move start after the repeated char

        # update last seen index of current character
        mapping[char] = end

        # update max_length for current window
        max_length = max(max_length, end - start + 1)

    return max_length


# test cases
print(longestSubstring("abcabcbb"))  # Expected output: 3 ("abc")
print(longestSubstring("bbbbb"))     # Expected output: 1 ("b")
print(longestSubstring("pwwkew"))    # Expected output: 3 ("wke")
